proposals/NNNN-clock-instant-date-duration.md

```diff
@@ -0,0 +1,498 @@
+
+#### Swift
+So given all of that, Swift can take this to another level of both accuracy of intent and ease of use than any of the other examples given. Following in the themes of other Swift APIs we can embrace the concept of progressive disclosure and leverage the existing frameworks that define time concepts.
+
+The given requirements are that we must have a way of expressing the frame of reference of time, this needs to be able to express a concept of now and a concept of waking up after a given instant has passed. Instants must be able to be compared among each other but are specific to the clock they were obtained. Instants also must be able to be advanced by a given duration or a distance between two instants must be able to emit a duration. Durations must be comparable and also must have some intrinsic unit of time that can suffice for broad application.
+
+It is worth noting that any extensions to Foundation, Dispatch or other frameworks beyond the swift standard library and concurrency library are not within the scope of this proposal and are under the prevue of those teams. This may or may not include additional ClockProtocol adoptions, additional functions that take the new types and changes in deprecations so any examples here are listed as illustrations of potential use cases and not to be considered as part of this proposal.
+
+##### Clock
+
+The base protocol for defining a clock requires two primitives; a way to wake up after a given instant, and a way to produce a concept of now.
+
+```swift
+public protocol ClockProtocol: Sendable {
+  associatedtype Instant: InstantProtocol
+  
+  var now: Instant { get }
+  
+  func sleep(until deadline: Instant) async throws
+}
+```
+
+This means that given an instant it is intrinsically linked to the clock; e.g. a monotonic instant is not meaningfully comparable to a wall clock instant. However as an ease of use concession the durations between two instants can be compared, however doing this across clocks is considered a programmer error unless handled very carefully. By making the protocol hierarchy just clocks and instants it means that we can easily express a compact form of a duration that is usable in all cases; particularly for APIs that might adopt Duration as a replacement to an existing type. 
+
+Clocks can then be used to measure a given amount of work. This means that clock should have the extensions to allow for the affordance of measuring workloads for metrics but also measure them for performance benchmarks. 
+
+```swift
+// measure with any clock
+public func measure<Clock: ClockProtocol>(clock: Clock, _ work: () async throws -> Void) reasync rethrows -> Duration
+
+// measure with a monotonic clock
+public func measure(_ work: () async throws -> Void) reasync rethrows -> Duration
+```
```

Comment:
Are they examples for how you can use `Clocks` to measure work, or are you proposing them as free-flow functions, not scoped anywhere? 


Also
> clock should have the extensions to allow for the affordance of measuring

If we think that clock should have these extensions, would it make sense to make them a requirement for `ClockProtocol`?